src/server/game/Server/Packets/ScenarioPackets.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+* Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
+*
+* This program is free software; you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the
+* Free Software Foundation; either version 2 of the License, or (at your
+* option) any later version.
+*
+* This program is distributed in the hope that it will be useful, but WITHOUT
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+* more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "ScenarioPackets.h"
+
+WorldPacket const* WorldPackets::Scenario::ScenarioState::Write()
+{
+    _worldPacket << int32(ScenarioId);
+    _worldPacket << int32(CurrentStep);
+    _worldPacket << int32(DifficultyId);
+    _worldPacket << int32(WaveCurrent);
+    _worldPacket << int32(WaveMax);
+    _worldPacket << int32(TimerDuration);
+    _worldPacket << int32(CriteriaProgress.size());
+    _worldPacket << int32(BonusObjectiveData.size());
+    _worldPacket << int32(StepProgressIndex.size());
+    _worldPacket << int32(UnkLegion.size()); // Unknown 
+
+    for (uint32 i = 0; i < StepProgressIndex.size(); ++i)
+    {
+        _worldPacket << uint32(StepProgressIndex[i].Id);
+    }
+
+    _worldPacket << uint8(ScenarioCompleted);
+
+    for (uint32 i = 0; i < CriteriaProgress.size(); ++i)
+    {
+        _worldPacket << uint32(CriteriaProgress[i].Id);
```

Comment:
this is very bad, this loop is shared with AllAchievementData - it needs to be refactored to expose
operator<<(WorldPackets::Achievement::CriteriaProgress const&) - and you should include it here